[13.0] Add module stock_move_common_dest_sync_location - alpha

grindtildeath opened the pull request:
Depends on: 
 - [x] #861

grindtildeath committed:
Add module stock_move_common_dest_sync_location

guewen commented on stock_move_common_dest_sync_location/models/stock_move_line.py:
```diff
@@ -0,0 +1,30 @@
+        if vals.get('location_dest_id') and not self.env.context.get('_sync_common_dest_location'):
+            related_moves = self.env['stock.move']
+            for move_line in self:
+                if not self.picking_type_sync_common_move_dest_location:
```
```suggestion
                if not move_line.picking_type_sync_common_move_dest_location:
```

Commit:
Fix warning in stock_move_common_dest

In this module: https://github.com/OCA/stock-logistics-warehouse/pull/808
A computed field uses "common_dest_move_ids" as dependency:

    @api.depends(
        "picking_type_id.display_completion_info",
        "move_lines.common_dest_move_ids.state",
    )
    def _compute_completion_info(self):

Which has the following effect when the ORM triggers changes:

odoo.osv.expression: Non-stored field stock.move.common_dest_move_ids cannot be searched.

Implement a search method to prevent this.

Commit:
apply new pre-commit

Commit:
Change sync of destination to happen from next move

First big change is that now, the 'sync' flag has to be set on the
"destination picking type", not the current one.

When we want to "regroup" moves from different transfer to the same
place for a packing, we want to activate this feature on the said
packing operation type.

Also, instead of changing the destination of move lines every time we change
a move line, we change it the first time a move is set to done to go
towards this location. We don't need to do it until the first line is
effectively moved.

Commit:
Add compatibility with stock_dynamic_routing

in stock_move_common_dest_sync_location

Some notes:

* syncing the destinations could not be done during _action_assign: if
we call _action_done on several moves at once with different locations,
we don't expect them to change of destination under the hood! So we sync
after _action_done
* For now, no code is needed in stock_dynamic_routing_common_dest_sync
to make the modules compatibles, but it contains tests with routing +
sync

jgrandguillaume reviewed (approved):
LGTM, though it'll need more extensive testing to see how it behave. I would be in favor of merging this as alpha because it works as expected. We just don't know now if the feature will be relevant in all context.

Commit:
Set the destination on the move as well

We want the move to be as precise as the move line once selected,
so for instance if we have:

Packing Post
Packing Post/Bay 1
Packing Post/Bay 2

When the first move is done using Packing Post/Bay 1, the moves
must be updated with Packing Post/Bay 1, because if we have unavailable
moves, we ensure that their move lines will be moved to Bay 1 once
they become available.

guewen reviewed (changes requested):
We shouldn't merge it. I'll soon propose a new solution to solve the issue.
I'll close this PR when I open the new one.

guewen commented:
The replacement for this feature is #907
